packages/phpstan-rules/src/Rules/RequireStringArgumentInConstructorRule.php

```diff
@@ -0,0 +1,155 @@
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Scalar\String_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\TypeWithClassName;
+use Symplify\Astral\Naming\SimpleNameResolver;
+use Symplify\PHPStanRules\Exception\ShouldNotHappenException;
+use Symplify\RuleDocGenerator\Contract\ConfigurableRuleInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\ConfiguredCodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * Useful for prefixed phar bulid, to keep original references to class un-prefixed
+ *
+ * @see \Symplify\PHPStanRules\Tests\Rules\RequireStringArgumentInConstructorRule\RequireStringArgumentInConstructorRuleTest
+ */
+final class RequireStringArgumentInConstructorRule extends AbstractSymplifyRule implements ConfigurableRuleInterface
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'Use quoted string in constructor "new %s()" argument on position %d instead of "::class. It prevent scoping of the class in building prefixed package.';
+
+    /**
+     * @var array<string, array<int>>
+     */
+    private $stringArgByType = [];
+
+    /**
+     * @var SimpleNameResolver
+     */
+    private $simpleNameResolver;
+
+    /**
+     * @param array<string, array<int>> $stringArgByType
+     */
+    public function __construct(SimpleNameResolver $simpleNameResolver, array $stringArgByType = [])
+    {
+        $this->stringArgByType = $stringArgByType;
+        $this->simpleNameResolver = $simpleNameResolver;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [New_::class];
+    }
+
+    /**
+     * @param New_ $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        $errorMessages = [];
+
+        foreach ($this->stringArgByType as $type => $positions) {
+            if (! $this->isNodeVarType($node, $scope, $type)) {
+                continue;
+            }
+
+            foreach ($node->args as $key => $arg) {
+                if ($this->shouldSkipArg($key, $positions, $arg)) {
+                    continue;
+                }
+
+                $errorMessages[] = sprintf(self::ERROR_MESSAGE, $type, $key);
+            }
+        }
+
+        return $errorMessages;
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(self::ERROR_MESSAGE, [
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+class AnotherClass
+{
+    public function run()
+    {
+        new SomeClass(YetAnotherClass:class);
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+class AnotherClass
+{
+    public function run()
+    {
+        new SomeClass('YetAnotherClass'');
+    }
+}
+CODE_SAMPLE
+                ,
+                [
+                    'stringArgByType' => [
+                        'SomeClass' => [0],
+                    ],
```

Comment:
The `0` looks like a position, but name is talking about `string`. I'm bit confused

Reply:
Copied from here https://github.com/symplify/symplify/blob/0b19f64f564e3b521a11dba9950cdcbe89fae199/packages/phpstan-rules/src/Rules/RequireStringArgumentInMethodCallRule.php#L118-L122

Reply:
I see. Both seems unreadable

Reply:
I fixed the one for this rule to `stringArgPositionsByType`, better?

Changing the other would be a BC break.

Reply:
This must be resolved too

Reply:
Fixed!